### ISSUE TITLE
Height of screenshots with cy.screenshot() in Chrome headless new is 79px smaller than viewportHeight config that is set

### ISSUE THREAD

**Issue by lgenzelis** [cypress-io/cypress#27260]
### Current behavior

Currently, calling `cy.screenshot()` during a test will use the viewport dimensions (specified in `cypress.config.js`) if the test is run using "headed" mode. However, when using headless mode:
- `viewportWidth` is respected
- `viewportHeight` is respected only if it's smaller than the "default". Otherwise, it's ignored.

That "default" height seems to depend on the browser being used. In cypress 12.14, the default was 720px for both chrome and firefox. However, in cypress 12.17, it's 720px for firefox, but 596px for chrome.

If you watch the video corresponding to the headless executions, sometimes you can even see, for a small fraction of a second, that the screen changes, removing all the Cypress UI elements, but also changing the viewport height.

https://github.com/cypress-io/cypress/assets/4582440/b748896a-1189-48f0-81a0-64336e2abfb4

A straightforward workaround for this is adding this line before capturing the screenshot:
```javascript
cy.get('html').then((html) => html.get(0).style.height = `${Cypress.config('viewportHeight')}px`);
```

 

### Desired behavior

Cypress should respect the specified `viewportHeight`.

### Test code to reproduce

https://github.com/lgenzelis/cypress-bug-screenshot-viewportheight-ignored

If you run this using `npm run cypress:open`, the screenshot is captured with the correct dimensions (specified in `cypress.config.js`). If run it using `npm run cypress:run`, though, the resulting screenshot is 400px wide (as expected), but only 596px tall. 

### Cypress Version

12.17.0

### Node version

v18.14.0

### Operating System

macos 13.4.1

**Comment by lgenzelis** [comment]
After more investigation, my "fix" is no real fix, it just makes the problem less obvious. The viewport is still taking the default height (e.g. 596px for chrome), but cypress default `'fullPage'` capture mode is making it look like it takes the correct viewport height. The problem only becomes apparent if there are any fixed elements in the UI (https://docs.cypress.io/api/commands/screenshot#Full-page-captures-and-fixedsticky-elements)

**Comment by lgenzelis** [comment]
Alright, after reading [this article](https://www.cypress.io/blog/2021/03/01/generate-high-resolution-videos-and-screenshots/#the-window-size-on-ci) (in particular, the distinction between viewport and window) I feel a little less flabbergasted. The only important thing to note, though, is that starting in cypress 12.15, chrome subtracts 124px from the window's height. I have no idea why, but [the release notes for 12.15](https://docs.cypress.io/guides/references/changelog#12-15-0) mention that "Chrome versions 112 and above will now be run in the headless mode that matches the headed browser implementation". So, there's definitely something there.

In a nutshell,  I don't know if this is a bug or not, but if you want to keep the same behavior as in before cypress 12.15, you need to add this to your `cypress.config.ts` (in `e2e.setupNodeEvents`):

```typescript
on('before:browser:launch', (browser, launchOptions) => {
  if (browser.name === 'chrome') {
    launchOptions.args.push(`--window-size=1280,${720 + 124}`);
  }
  return launchOptions;
});
```

(note the magical `+124` part 😅

**Comment by lmiller1990** [comment]
I could be misunderstanding or doing it wrong, but I could not reproduce this. I tried making a page that is 738px high (4x 180px + 18px) and used Cypress 12.14, then Cypress 12.17. Both screenshots are the same - 738px high:

https://github.com/lmiller1990/cypress-test-tiny/compare/master...lmiller1990:cypress-test-tiny:issue-27260?expand=1

```
$ file before.png                                               (git)-[master]-
before.png: PNG image data, 1000 x 738, 8-bit/color RGBA, non-interlaced
$ file after.png                                                (git)-[master]-
after.png: PNG image data, 1000 x 738, 8-bit/color RGBA, non-interlaced
```

I tried with `capture: 'viewport'` and did see a 4px difference - weird. It's not the 124px you described, though. 🤔 

Would be nice to find out what exactly changed, this could be a huge rabbit hole. Any recommendation for moving forward? We are using Percy for visual regression and none of our tests (500+ screenshots) had any changes... strange.

**Comment by lgenzelis** [comment]
That is really strange, @lmiller1990 . Adding that `+ 124` consistently brought the height back to `720` for me, as in before cypress 12.15. I reproduced it both locally (in macos) and in github actions running with ubuntu-latest. Two questions:
- are you sure you're using chrome (in headless mode)? This doesn't happen with e.g. firefox (and I didn't try electron)
- if so, are you using chrome 114? The change I refer to in the changelog for cypress 12.15 mentions "Chrome versions 112 and above"

**Comment by lmiller1990** [comment]
Default is `--headless`, so definitely headless. Chrome 114.

![image](https://github.com/cypress-io/cypress/assets/19196536/04425e39-0a09-4499-8a79-5b4ef0b95feb)

Glad you found a fix, it would be nice if we could reproduce and narrow the issue down.

**Comment by jmeleropager** [comment]
I ran into same issue and after digging deeper I believe I found the source of the issue: The new headless mode of Chrome is "rendering" Chrome's UI alongside the viewport.

I've done a quick test, running these commands:

```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --headless=old --window-size=1280,800 --screenshot https://developer.chrome.com/ 

/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --headless=new --window-size=1280,800 --screenshot https://developer.chrome.com/ 
```

### Results

Old headless mode (image resolution 2560x1600)
![screenshot-old](https://github.com/cypress-io/cypress/assets/87527748/59f83011-7e43-4cf0-b52c-7fa24f59e947)

New headless mode (image resolution 2560 × 1442)
![screenshot-new](https://github.com/cypress-io/cypress/assets/87527748/c9b03582-0137-40a1-8270-ac24d5817990)

From previous images I can drop 2 conclusions:
- Both modes are aware of underlying SO/hardware, because taking screenshot of 1280x800 window size resulted in images of double resolution, I guess it's because I'm using a Mac with retina display.
- New headless mode image has a lower vertical resolution: 1442 instead of 1600, 158px less.

Trying to explain the difference in last point, I took a screenshot of the browser and measured chrome's UI

<img width="729" alt="Captura de pantalla 2023-07-12 a la(s) 4 21 58 p m  1" src="https://github.com/cypress-io/cypress/assets/87527748/334df1d6-abaf-4f3a-ac27-be795b2f5092">

And not for my surprise, yes, Google Chrome UI is **158px height** (actually is 79px height, but because of retina display the image has 158px).

So yes, it seems new Google Chrome "headless" mode is actually rendering the UI or at least is making space for it.

<img src="https://media0.giphy.com/media/15BuyagtKucHm/giphy.gif?cid=6104955ee8rtw7924abz65wtge5ljqdvfybdh99djgc9zbfy&ep=v1_gifs_translate&rid=giphy.gif&ct=g" width="480" alt="">

**Comment by lgenzelis** [comment]
I've just run those two commands in my computer as well and got the exact same results as @jmeleropager 🤯 So, the problem is caused indeed by Cypress 12.15 changing to `--headless=new`.

**Comment by lmiller1990** [comment]
To clarify what headless: new is, see https://developer.chrome.com/articles/new-headless/. 

As for this discovery; is the proposed fix to simply add a hard-coded value of 124 to the viewport height during a chrome headless=new screenshot? If this really fixes the issue and Cypress continues to work how you'd expect, this seems like a reasonable proposed fix to me.

**Comment by yktoo** [comment]
I suspect it's also related to #27291. The screenshotting mechanics seems totally broken:
* the viewport settings are ignored
* ~~some screenshots contain Cypress sidebar, some inexplicably don't~~ turns out the sidebar is only present when registering a failure, not when calling `cy.screenshot()`
* the screenshots without the sidebar contain garbage, seemingly leftovers from previous screenshots

**Comment by lmiller1990** [comment]
@yktoo is this a general bug report of specific to chrome 114+ and the new headless mode (which seems to be the general consensus here in this issue).

If there's a separate bug, we should file a separate issue - if something is totally broken, this is a high priority (although all the screenshots in my projects look okay, it's definitely possible something regressed, although my understanding is nothing changed around there lately).

**Comment by yktoo** [comment]
@lmiller1990 Well it depends of course on the definition of "totally," but to me the combination of this one and #27291 renders screenshots pretty much unusable. The good news is failure screenshots are not affected (so it seems at least).

I believe these two bugs cover the existing problems.

**Comment by lmiller1990** [comment]
#27291 looks like a huge problem - I don't see this on my apps, but we will look into what's happening there - a core feature like this should work perfectly (and our tests say it does... maybe we missed some cases).

**Comment by 4knort** [comment]
```
on('before:browser:launch', (browser, launchOptions) => {
  if (browser.name === 'chrome') {
    launchOptions.args.push(`--window-size=1280,${720 + 124}`);
  }
  return launchOptions;
});
```

this workaround doesn't work if we screenshot elements, not full page. Elements are smaller than they need to be, if our website is flexible.

**Comment by MelvinBaaijAviva** [comment]
I am also experiencing issues with screenshots (viewportwidth / viewportheight are not respected and or screenshot is made with a zoomlevel). Very annoying when using visual testing (for example with cypress-image-diff-js), especially on specific elements. 
I hope you can have a look on this one and fix it @lmiller1990 

@4knort did you already find a workaround or solution for your problem?

**Comment by hanna-becker** [comment]
Any updates here? We've been stuck with v12.14.0 because of this issue (upgrading breaks our snapshot tests).

**Comment by vaab** [comment]
Same here, recently upgraded to cypress `13.5.1` over a cypress `12.11.0`. And noticed that our tour of the app, used to make screenshots for mobile app stores, produced different sized screenshots. Only the height is affected. And yes, I'm using chrome 114, and could reproduce this on macosx and linux box on our CI, also:

The report at the end of the tests
```
  -  /.../.cypress/screenshots/spec.cy.ts/login.png                  (1242x1857)
  -  /.../.cypress/screenshots/spec.cy.ts/dashboard.png              (1242x2208)
  -  /.../.cypress/screenshots/spec.cy.ts/recipients.png             (1242x2208)
  -  /.../.cypress/screenshots/spec.cy.ts/pay.png                    (1242x2208)
  -  /.../.cypress/screenshots/spec.cy.ts/top-up.png                 (1242x2208)
  -  /.../.cypress/screenshots/spec.cy.ts/map.png                    (1242x1857)
```

Expected size: 1242x2208.

Two of them are not correctly sized (first and last), but actually all of them are screwed: on the ones that seem to have the right height, the actual bit below 1857 and down to the bottom is filled with garbage that seem to be previous screenshots or some cache of what was rendered before.

This seems pretty random as it occurs on the same run. Here, with my settings, the padding could be 117 (x3 which the scaleFactor). If this works for this resolution it feels clunky. And I tried with success the forcing of `--headless=old`. Which worked flawlessly.
So for me, this is the best workaround: you add in your `cypress.config.ts`:

```
on('before:browser:launch', (browser, launchOptions) => {
  if (browser.name === 'chrome') {
    launchOptions.args.push(`--headless=old`);
  }
  return launchOptions;
});
```

**Comment by emilyrohrbough** [comment]
There are two things going on here:

## Wont Fix - Browser Behavior

Cypress updated to use Chrome's `headless=new` flag when testing. Chrome introduced this flag to unify how the headed and headless browsers were rendered for a more accurate headless browser implementation. This reduction to the viewport height available to render browser content is outside of Cypress' control and could change in any Chrome version. With the `headless=new` flag, this is the correct behavior. The introduction of the `headless=new` flag likely should have been apart of v13.

## Bug - Capture `fullPage` is incorrectly capturing as `viewport`

As reported, the default screenshot capture method of `fullPage` should be respecting the configured `viewportHeight`. 
The current screenshot behavior is acting as `capture: viewport` instead of `capture: fullPage`.

The Chrome change impacting the viewport height, should ONLY be impacting `cy.screenshot({ capture: 'viewport' })` / `Cypress.screenshot.defaults({ capture: 'viewport' })`.  However as observed in your reproduction, the `fullPage` screenshot is not actually capturing a the full AUT, but what is visible within the viewport when the Cypress runner UI is removed. 

The `cy.screenshot` command is correctly taking 1+ screenshots to account for the `viewportHeight` value, however, our styles are incorrect so they window is not scrollable, so when the scroll logic executes to capture content outside of the viewport, it does not scroll.

I quickly tested Cypress v9 and its was also broken, so this been an issue for a few versions. Fixing this will have a fairly large impact for anyone utilizing screenshots for visual regression. cc// @jennifer-shehane 

## Varying Heights per Run
@vaab Per https://github.com/cypress-io/cypress/issues/27260#issuecomment-1847500569, I have been unable to reproduce this behavior. Please open a new issue specifically around this issue you are reporting.

**Comment by jennifer-shehane** [comment]
After a thorough evaluation of this described issue, Cypress has decided that the new behavior from Chrome headless=new in this case is something that we will not be changing our behavior around. 

With Chrome updating their headless browser to be the same codebase as the headed browser, this introduced Chrome's 'chrome' and UI like info bars taking up space within the window. This height is not under our control and it is able to change based on Chrome's implementation.

I did find this issue on Chromium's side, where someone was saying this behavior should be the same as headless=old and Chromium should change it. I'm not sure if they intend to act on this though: https://issues.chromium.org/issues/40275252

We could arbitrarily add some height calculations, but this won't solve anything since this height is able to change based on Chrome's implementation. 

We'll be closing this issue.

**Comment by lgenzelis** [comment]
I had forgotten about this issue. But for anyone running into it, we just fixed by passing a window-size big enough to account for both your app `viewportHeight` and for whatever extra chrome decides to render. The sweet thing is that it just needs to be big *enough*, it doesn't need to be "just right". So, for example, in our project's `cypress.config.ts` we have:

```ts
const viewportWidth = 1200;
const viewportHeight = 720;

export default defineConfig({
  viewportWidth,
  viewportHeight,
  e2e: {
    setupNodeEvents(on, config) {
      // details about this API: https://docs.cypress.io/api/plugins/browser-launch-api
      on('before:browser:launch', (browser, launchOptions) => {
        if (browser.name === 'chrome' && browser.isHeadless) {
          // `launchOptions.args` are the arguments that will be passed to the browser when Cypress launches it.
          // we are setting a fixed height of 1000px for the browser window, since that's big enough for our desired viewport
          // height of 720px.
          launchOptions.args.push(`--window-size=${viewportWidth},1000`);
        }
        return launchOptions;
      });

      return config;
    },
  },
});
```
Meaning, we want a `viewportHeight` of `720px`, so we just set a `--window-size=1200,1000` and it works (1000 is "big enough" for the 720 target)

**Comment by crfrolik** [comment]
The workaround of using `--headless=old` as a Chromium launch option broke our tests with Chrome 132 (Chrome 131 and earlier still work).

We get errors like:

```
Cypress failed to make a connection to the Chrome DevTools Protocol after retrying for 50 seconds.

This usually indicates there was a problem opening the Chrome browser.
```

Looks like another workaround might be needed.

**Comment by crfrolik** [comment]
https://developer.chrome.com/blog/removing-headless-old-from-chrome

**Comment by jennifer-shehane** [comment]
The width and the height of screenshots is variable across major versions of all of the browsers - we've seen this over many versions. You'll need to work with the implementation that the browser decides, as we're not going to update our logic to workaround every time they change what they include in the width/height.
